llama-index-core/llama_index/core/storage/docstore/keyval_docstore.py

```diff
@@ -374,90 +374,92 @@ async def adocument_exists(self, doc_id: str) -> bool:
         """Check if document exists."""
         return await self._kvstore.aget(doc_id, self._node_collection) is not None
 
-    def _remove_ref_doc_node(self, doc_id: str) -> None:
-        """Helper function to remove node doc_id from ref_doc_collection."""
+    def _get_ref_doc_id(self, doc_id: str) -> Optional[str]:
+        """Helper function to get ref_doc_info for a given doc_id."""
         metadata = self._kvstore.get(doc_id, collection=self._metadata_collection)
         if metadata is None:
-            return
-
-        ref_doc_id = metadata.get("ref_doc_id", None)
-
-        if ref_doc_id is None:
-            return
-
-        ref_doc_info = self._kvstore.get(
-            ref_doc_id, collection=self._ref_doc_collection
-        )
-
-        if ref_doc_info is not None:
-            ref_doc_obj = RefDocInfo(**ref_doc_info)
-
-            ref_doc_obj.node_ids.remove(doc_id)
-
-            # delete ref_doc from collection if it has no more doc_ids
-            if len(ref_doc_obj.node_ids) > 0:
-                self._kvstore.put(
-                    ref_doc_id,
-                    ref_doc_obj.to_dict(),
-                    collection=self._ref_doc_collection,
-                )
+            return None
 
-            self._kvstore.delete(ref_doc_id, collection=self._metadata_collection)
+        return metadata.get("ref_doc_id", None)
 
-    async def _aremove_ref_doc_node(self, doc_id: str) -> None:
-        """Helper function to remove node doc_id from ref_doc_collection."""
+    async def _aget_ref_doc_id(self, doc_id: str) -> Optional[str]:
+        """Helper function to get ref_doc_info for a given doc_id."""
         metadata = await self._kvstore.aget(
             doc_id, collection=self._metadata_collection
         )
         if metadata is None:
-            return
+            return None
 
-        ref_doc_id = metadata.get("ref_doc_id", None)
+        return metadata.get("ref_doc_id", None)
 
+    def _remove_from_ref_doc_node(self, doc_id: str) -> None:
+        """
+        Helper function to remove node doc_id from ref_doc_collection.
+        If ref_doc has no more doc_ids, delete it from the collection.
+        """
+        ref_doc_id = self._get_ref_doc_id(doc_id)
         if ref_doc_id is None:
             return
+        ref_doc_info = self._kvstore.get(
+            ref_doc_id, collection=self._ref_doc_collection
+        )
+        if ref_doc_info is None:
+            return
+        ref_doc_obj = RefDocInfo(**ref_doc_info)
+        ref_doc_obj.node_ids.remove(doc_id)
```

Comment:
as a sanity check, should we make sure doc_id is in node_ids (I've seen this error out, somehow, before?)

Reply:
Or maybe we should be raising an error if it gets to this state. Thoughts?